src/sqlparser/src/parser_v2/data_type.rs

```diff
@@ -0,0 +1,225 @@
+                    Ok(())
+                }
+            },
+        )),
+    ))
+    .context(StrContext::Label("data_type"))
+    .parse_next(input)
+}
+
+/// Data type parsing with stateful stream.
+fn data_type_stateful<S>(input: &mut StatefulStream<S>) -> PResult<DataType>
+where
+    S: TokenStream,
+{
+    repeat(0.., (Token::LBracket, cut_err(Token::RBracket)))
+        .fold1(data_type_stateful_inner, |mut acc, _| {
+            acc = DataType::Array(Box::new(acc));
+            acc
+        })
+        .parse_next(input)
+}
+
+/// Consume a data type except [`DataType::Array`].
+fn data_type_stateful_inner<S>(input: &mut StatefulStream<S>) -> PResult<DataType>
+where
+    S: TokenStream,
+{
+    let with_time_zone = || {
+        opt(alt((
+            (Keyword::WITH, Keyword::TIME, Keyword::ZONE).value(true),
+            (Keyword::WITHOUT, Keyword::TIME, Keyword::ZONE).value(false),
+        )))
+        .map(|x| x.unwrap_or(false))
+    };
+
+    let precision_and_scale = || {
+        opt(delimited(
+            Token::LParen,
+            (literal_uint, opt(preceded(Token::Comma, literal_uint))),
+            Token::RParen,
+        ))
+        .map(|p| match p {
+            Some((x, y)) => (Some(x), y),
+            None => (None, None),
+        })
+    };
+
+    let keywords = dispatch! {keyword;
+        Keyword::BOOLEAN | Keyword::BOOL => empty.value(DataType::Boolean),
+        Keyword::FLOAT => opt(precision_in_range(1..54)).map(DataType::Float),
+        Keyword::REAL => empty.value(DataType::Real),
+        Keyword::DOUBLE => opt(Keyword::PRECISION).value(DataType::Double),
+        Keyword::SMALLINT => empty.value(DataType::SmallInt),
+        Keyword::INT | Keyword::INTEGER => empty.value(DataType::Int),
+        Keyword::BIGINT => empty.value(DataType::BigInt),
+        Keyword::STRING | Keyword::VARCHAR => empty.value(DataType::Varchar),
+        Keyword::CHAR | Keyword::CHARACTER => dispatch! {keyword;
+            Keyword::VARYING => empty.value(DataType::Varchar),
+            _ => opt(precision_in_range(..)).map(DataType::Char),
+        },
+        Keyword::UUID => empty.value(DataType::Uuid),
+        Keyword::DATE => empty.value(DataType::Date),
+        Keyword::TIMESTAMP => with_time_zone().map(DataType::Timestamp),
+        Keyword::TIME => with_time_zone().map(DataType::Time),
+        // TODO: Support complex interval type parsing.
+        Keyword::INTERVAL => empty.value(DataType::Interval),
+        Keyword::REGCLASS => empty.value(DataType::Regclass),
+        Keyword::REGPROC => empty.value(DataType::Regproc),
+        Keyword::TEXT => empty.value(DataType::Text),
+        Keyword::STRUCT => cut_err(struct_data_type).map(DataType::Struct),
+        Keyword::BYTEA => empty.value(DataType::Bytea),
+        Keyword::NUMERIC | Keyword::DECIMAL | Keyword::DEC => cut_err(precision_and_scale()).map(|(precision, scale)| {
+            DataType::Decimal(precision, scale)
+        }),
+        _ => fail,
```

Comment:
Hitting this branch means that we've successfully got a keyword but it does not match any data type. Should we `cut_err` here?